### PR TITLE
Test strategy

### PR DESCRIPTION
**Pull request by bobturneruk**
Fixes *NA*.

### Description
Lays the foundations for adding **unit** and **regression** tests to `pykale`, and reporting the coverage to **codecov**.

A [folder structure is proposed](https://github.com/pykale/pykale/tree/test-strategy/tests) that mimics that of the pykale module. Each `.py` file in `pykale` is given a `tests_<filename>.py` which is to contain unit tests for all functionality in its companion file. 

Regression tests are given a [separate folder](https://github.com/pykale/pykale/tree/test-strategy/tests/regression). I propose these are based on [existing examples](https://github.com/pykale/pykale/tree/test-strategy/examples). 

A [folder has been created (with an empty `.csv` file)](https://github.com/pykale/pykale/tree/test-strategy/tests/test_data) to hold data needed for testing. Depending on amount and format of required data this will need to be revisited. For example a few baseline `.csv` files would be fine, but loads of images or video may be better held elsewhere e.g. linked via a DOI.

The following must be executed by CI post tests to report data to codecov:

```
coverage xml
bash <(curl -s https://codecov.io/bash)
```

### Status
**Work in progress**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).

### CODE REVIEW

File: setup.py
Comment (willfurnass) [inline setup.py:26]: Loosely related: separate out `extra_` dev dependencies and dependencies for tutorials/examples?

Comment (bobturneruk) [reply setup.py:26]: I think it's been decided that this is OK for now, but there's some ongoing thinking.

**Comment by bobturneruk** [comment]
@haipinglu do you have any views on what level of "problem" in `numpy` floating point operations should cause tests to fail? If the defaults here are OK, that's fine. Otherwise I'll figure a way to set these consistently over the test suite.

https://numpy.org/doc/stable/reference/generated/numpy.seterr.html#numpy.seterr

**Comment by haipinglu** [comment]
> @haipinglu do you have any views on what level of "problem" in `numpy` floating point operations should cause tests to fail? If the defaults here are OK, that's fine. Otherwise I'll figure a way to set these consistently over the test suite.
> 
> https://numpy.org/doc/stable/reference/generated/numpy.seterr.html#numpy.seterr

Sorry I do not have much knowledge on this issue. My suggestion is that if urgent, email the whole team (pykale-group@shef.....), otherwise, put a card on this in the project board so that we could discuss in our next PyKale meeting. Many thanks!

**Comment by bobturneruk** [comment]
Thanks @haipinglu - I don't have a strong opinion on this, so I will just make a note of it in my draft.

**Comment by bobturneruk** [comment]
@haipinglu Could the build and test GitHub actions workflows be consolidated? It seems like they are quite similar.

**Comment by haipinglu** [comment]
> @haipinglu Could the build and test GitHub actions workflows be consolidated? It seems like they are quite similar.

That's right. Great proposal! Put all workflows that need installation (which takes time) under one workflow will be more compact and efficient, saving the overall CI time.

How shall we proceed? Do the "merge" of these two workflows here or create a new dedicated branch?

**Comment by bobturneruk** [comment]
I'll work a proposal into this PR. You may ultimately also want a "release" action that automates releases when a new tag is made.

**Comment by haipinglu** [comment]
> I'll work a proposal into this PR. You may ultimately also want a "release" action that automates releases when a new tag is made.

That's right. Great suggestion. I've bookmarked two related actions for this purpose but further dig is needed:

1. https://github.com/marketplace/actions/pypi-publish 
2. https://github.com/actions/create-release  

Maybe I should consider using workflow for the next release to automate the process. I won't target the next release before the Chinese New Year.

Regarding the conda release, I did not find it useful at the moment (2 downloads in conda vs 200 in PyPI). Since PyPI is still the primary channel, shall we remove it to simplify the file structure and our process?

**Comment by bobturneruk** [comment]
Release automation - depends on time we have to work on it and your release schedule. It's mostly nice as it codifies the release process. 

Re: conda I think that's fine. I'm considering dropping conda cloud for another project.

**Comment by willfurnass** [comment]
@bobturneruk Should it be useful I had a stab at release automation for MuMoT: https://github.com/DiODeProject/MuMoT/blob/master/.github/workflows/test-and-release.yml

**Comment by bobturneruk** [comment]
Thanks @willfurnass - we did something similar in Travis https://github.com/ScottishCovidResponse/data_pipeline_api/blob/master/.travis.yml GitHub actions better.

File: tests/utils/test_seed.py
Comment (drj11) [inline tests/utils/test_seed.py:19]: Why use almost_equal? For this situation, you really do want them to test that the random number is exactly the same, otherwise we're not going to get the same outputs.

**Review by drj11 — COMMENTED** [review]
not really a full review, but it seems github makes me fill this in anyway

File: tests/utils/test_seed.py
Comment (bobturneruk) [reply tests/utils/test_seed.py]: Thanks @drj11, that's better.

**Comment by bobturneruk** [comment]
Hi all! I think this is about ready for a broader review. We need to think about whether we want to set up codecov at this time (as we have but 1% coverage). codecov link is currently dead.

**Comment by bobturneruk** [comment]
@haipinglu - please tag other to review as you see fit.

File: CONTRIBUTING.md
Comment (haipinglu) [inline CONTRIBUTING.md:138]: Many (including me before Googling) may not know what are unit/regression tests. It will be good to add top references for these specific technical terms. The link to realpython is a good reference while it also has lots of ads and limited viewing space, which is not pleasant and efficient (at least to me). 

Will those from the official python docs be good (for beginners)? 

For example, are the following (from the official docs) good and appropriate references?
unit tests: https://docs.python.org/3/library/unittest.html#module-unittest
regression tests: https://docs.python.org/3/library/test.html

Will doctest be useful? https://docs.python.org/3/library/doctest.html

The above references are the official development tools docs: from https://docs.python.org/3/library/development.html

File: CONTRIBUTING.md
Comment (haipinglu) [inline CONTRIBUTING.md:142]: add ref? https://docs.pytest.org/en/stable/

File: CONTRIBUTING.md
Comment (haipinglu) [inline CONTRIBUTING.md:161]: I think `DataFrame` should be one word,  as in Pandas and Spark/PySpark. documentations. 
The same for the following.

File: CONTRIBUTING.md
Comment (haipinglu) [inline CONTRIBUTING.md:165]: floting-->floating

File: CONTRIBUTING.md
Comment (haipinglu) [inline CONTRIBUTING.md:194]: An example or reference will further clarify this issue.

**Review by haipinglu — COMMENTED** [review]
I have gone through all changes. Great efforts. Many thanks Bob. This could be a significant step in ensuring the **reproducibility** of our code as well, which is currently troubling @XianyuanLiu 

I've made some minor comments. Many other important issues have arisen as well, with some captured in the project board cards.

If I understand it correctly, this is not yet ready to merge yet and we need to discuss the steps moving towards a merge so we could start solid integration. Personally, I prefer to merge as long as it can be merged so that we could at least have some kind of tests in place (even just the seed.py, which is important for _reproducibility_). We could discuss on Thursday.

Regarding your question "whether we want to set up codecov at this time (as we have but 1% coverage)", I think we should have it but monitor it separately, e.g. as a card on the project board, instead of putting a badge there (i.e. comment it until a decent percentage to show visitors). 

This is another new territory for me and it will take me some time to pick up, but I am a keen learner :-)

My todos:
- Release automation 
- Remove conda
- Add more options for `extras`, at least a `dev`

File: README.md
Comment (haipinglu) [inline README.md:7]: No longer valid after deleting build workflow.

File: setup.py
Comment (haipinglu) [reply setup.py:26]: This has been addressed in https://github.com/pykale/pykale/tree/auto-release-and-setup and PR #60

Comment (bobturneruk) [reply setup.py:26]: Great, thanks. I'll have a look at that PR as time allows.

**Comment by RaivoKoot** [comment]
Great additions. Thank you! I very much like the detailed explanations in the README and the CONTRIBUTING.

**Comment by RaivoKoot** [comment]
Unsure whether I should give an "approved" review because, I think, it says that would start an automatic merge, and I can see in the comments above that you were unsure about merging. Let me know if I should. Am ready to approve.

**Comment by bobturneruk** [comment]
@RaivoKoot - good spot. I'd leave it for now. I'm not 100% sure automatic merge is a good idea. It's fast, but when there's "disagreement" between reviewers it's tricky. Policy can state it's the person who makes the PR who should merge. Slower, but the important thing, I think, it that it's clear who's responsibility it is to merge.

**Comment by haipinglu** [comment]
@bobturneruk @RaivoKoot There seems to be some misconception. The auto-merge is an **option** that needs to be enabled for each PR individually rather than an _always-on_ feature. For this PR, the auto-merge is NOT enabled until someone click the enable auto-merge so an approval will NOT trigger an auto-merge unless this option is activated. We still have full control and the one who enables auto-merge should be responsible, which is similar to manual merge where the one who merges it should be sure it is ready for merge. I will clarify this point in #58 .

**Comment by haipinglu** [comment]
> @RaivoKoot - good spot. I'd leave it for now. I'm not 100% sure automatic merge is a good idea. It's fast, but when there's "disagreement" between reviewers it's tricky. Policy can state it's the person who makes the PR who should merge. Slower, but the important thing, I think, it that it's clear who's responsibility it is to merge.

See my clarification at https://github.com/pykale/pykale/pull/58/commits/7b7b09b6a246492955f51dd8f4ed124c9241c50b

**Comment by RaivoKoot** [comment]
> @bobturneruk @RaivoKoot There seems to be some misconception. The auto-merge is an **option** that needs to be enabled for each PR individually rather than an _always-on_ feature. For this PR, the auto-merge is NOT enabled until someone click the enable auto-merge so an approval will NOT trigger an auto-merge unless this option is activated. We still have full control and the one who enables auto-merge should be responsible, which is similar to manual merge where the one who merges it should be sure it is ready for merge. I will clarify this point in #58 .

Ahh thank you for clarifying. That sounds good. Approved from me in that case.

**Review by RaivoKoot — APPROVED** [review]

File: CONTRIBUTING.md
Comment (haipinglu) [reply CONTRIBUTING.md:138]: https://github.com/ScottishCovidResponse/simple_network_sim/tree/master/tests

Comment (bobturneruk) [reply CONTRIBUTING.md:138]: Add links to official pytest docs.
Add links to examples https://github.com/ScottishCovidResponse/simple_network_sim
Add links to other tutorials etc.

**Comment by haipinglu** [comment]
@bobturneruk I have resolved some minor issues (e.g. typo) and made some minor updates. The remaining issues will be best resolved by you. The next merge should be fine with or without merging this PR since the current version of guidance has mentioned tests and there will be a learning curve for new comers.

I added a [codecov card](https://github.com/pykale/pykale/projects/9#card-55216673) on the project board for us to monitor the codecov status. After reaching a decent coverage, we could add this badge back.

**Review by sz144 — APPROVED** [review]

File: CONTRIBUTING.md
Comment (haipinglu) [reply CONTRIBUTING.md:139]: "All new code must be covered by unit tests, and regression tests where appropriate.": This would be great of course but it could be hard to achieve in the short term (for my part, at least until I have learned how to write one). We need to see what will be a feasible path and strike a right balance.

**Comment by bobturneruk** [comment]
Hi @haipinglu - can you add a "testing" topic to discussions, please? I'll monitor this and try to provide support. See:

https://github.com/pykale/pykale/pull/57/commits/1a90ca892a96216d1e511649ca7ce9bd057fab72

**Comment by haipinglu** [comment]
> Hi @haipinglu - can you add a "testing" topic to discussions, please? I'll monitor this and try to provide support. See:
> 
> [1a90ca8](https://github.com/pykale/pykale/commit/1a90ca892a96216d1e511649ca7ce9bd057fab72)

I've created a new [category "Testing"](https://github.com/pykale/pykale/discussions/categories/testing) and updated the link https://github.com/pykale/pykale/pull/57/commits/e4951b48ae81386e98c3d95cbe7f380b5c036578. If I misunderstood, please let me know.

Thanks a lot for the update. I will study in more details and look forward to discussions on Thursday.

**Comment by haipinglu** [comment]
@bobturneruk Have you finished your update (it seems so)? How shall we proceed from here? Merge and then discuss on Thursday or discuss on Thursday before merge?

**Comment by bobturneruk** [comment]
Hi @haipinglu - I've asked my colleagues to double check my definitions, and was waiting for that, but I think this can be merged and subsequently discussed / corrected based on further feedback.

**Comment by bobturneruk** [comment]
OK - I've had some feedback and it's fine. I'd recommend merging.

One point - this is way bigger than I would normally like a PR to be - it's not ended up being exemplary GitHub practise!

**Comment by haipinglu** [comment]
> OK - I've had some feedback and it's fine. I'd recommend merging.
> 
> One point - this is way bigger than I would normally like a PR to be - it's not ended up being exemplary GitHub practise!

Will do. Thank you! We'll strike a better balance next time.

**Review by haipinglu — APPROVED** [review]

**Pull request merged**
